pkg/extract/update_strategy.go

```diff
@@ -0,0 +1,90 @@
+package extract
+
+import (
+	"reflect"
+
+	"golang.stackrox.io/kube-linter/pkg/k8sutil"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// UpdateStrategy will extract the UpdateStrategy struct from a provided
+// object if it exists
+func UpdateStrategy(obj k8sutil.Object) (interface{}, bool) {
```

Comment:
Instead of depending on the caller to chain these functions in the appropriate way, what do you think of just having a function like:
```go
func UpdateStrategyFields(obj k8sutil.Object) (type string, maxUnavailable, maxSurge *intstr.IntOrString)
```
and having the function just return these? You can return `type` as `""` if not found, and the others as `nil`. The caller code then becomes much simpler, and all the functions you have here can be unexported.

Reply:
I originally thought about doing that and headed down that route, but then it makes the extract functions very specific.  I didn't put the implementation in, but StatefulSets have an UpdateStrategy field, but only have it filled with a partition field.  Likewise, DeploymentConfigs have a lot more fields than maxUnavailable and maxSurge.  I ended up with this path to make it easier to reuse the functions for later checks if anything came up.

I suppose in the suggestion you make, the path for anything that wants to extract additional information is to create new extraction functions that use some of the non-exported functions.  Is that the path you'd prefer?

Reply:
No, my suggestion was a lot simpler. Basically, right now, you're doing all this in the `template`:

[Call the following code snippet 1]
```go
				strategy, found := extract.UpdateStrategy(object.K8sObject)
				strategyType, found := extract.TypeFromUpdateStrategy(strategy)
				rollingUpdate, found := extract.RollingUpdateFromUpdateStrategy(strategy)
				maxUnavailable, found := extract.MaxUnavailableFromRollingUpdate(rollingUpdate)
				maxSurge, found := extract.MaxSurgeFromRollingUpdate(rollingUpdate)
```

This requires quite a bit of work from the caller, and requires the caller to understand the internals of this code. Instead of this, I would propose that we have a function in the `extract` package like:
```go
func UpdateStrategyFields(obj k8sutil.Object) (type string, maxUnavailable, maxSurge *intstr.IntOrString)
```
internally, this function would just be making the series of calls that I listed above (in code snippet 1). The difference now is that the `template` gets all this in one call.

For better readability and extensibility, we can also change my proposed function's signature so that instead of returning a lot of values, we define a custom struct like `UpdateStrategyData` or whatever, which has all these fields, and then fill those. IMO, that will provide a much cleaner interface for the caller.

Does that make sense?

Reply:
That is what I understood you to mean, save the custom struct bit.  You've described the first pass I took at this, but in the end I ended up not liking what the caller had to do with a custom struct to handle the variable results/scenarios iirc.  I've since changed logic of things quite a bit so maybe some of those concerns are gone.

I can refactor this way again and see if the structure of things is better now that the template logic is worked out better.

Reply:
Done